### PR TITLE
gh-126914: Store the Preallocated Thread State's Pointer in a PyInterpreterState Field

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This approach eliminates the originally reported race.  It also gets rid of the deadlock reported in gh-96071, so we can remove the workaround added then.

* Issue: gh-126914

### CODE REVIEW

**Comment by ZeroIntensity** [comment]
I'll get to reviewing in an hour or so. At a glance:

- Should we backport this? I'm concerned we might make debuggers unhappy on 3.13
- We should consider a lock-free algorithm for the freelist to help ease the contention if a lot of subintepreters are trying to get created.
- If I understand correctly, this only applies to the preallocated initial thread. Why not make all thread states participate in the freelist?

File: Python/pystate.c
Comment (ZeroIntensity) [inline Python/pystate.c:1412]: Needs a NULL check

File: Python/pystate.c
Comment (ZeroIntensity) [inline Python/pystate.c]: Regarding trying to do this without a lock, we could do something like:

```c
tstate = _Py_atomic_exchange_ptr(&interp->threads.freelist.head, NULL);
```

I'm not totally sure what the performance benefit would be, but I *think* it should be a little better than a lock.

File: Python/pystate.c
Comment (ZeroIntensity) [inline Python/pystate.c]: Doing this part without a lock would be a little trickier, it's probably something like:

```c
PyThreadState *top = _Py_atomic_load_ptr_relaxed(&interp->threads.freelist.head);
int result;
while ((result = _Py_atomic_compare_exchange_ptr(&interp->threads.freelist.head, &top, tstate) == 0)
{
    top = _Py_atomic_load_ptr_relaxed(&interp->threads.freelist.head);
}
tstate->base.next = top;
```

File: Python/pystate.c
Comment (ZeroIntensity) [inline Python/pystate.c:1429]: `alloc_threadstate` already resets the thread state, it's probably fine to keep it in the freelist while still initialized (in fact, that will hopefully be helpful in debugging).

File: Lib/test/test_interpreters/test_stress.py
Comment (ZeroIntensity) [inline Lib/test/test_interpreters/test_stress.py:36]: I'm not sure if `requires_resource('cpu')` can pass without working threading, but adding that seemed to help fix an issue I was having with threading in CI the other day (see GH-126849)
```suggestion
    @support.requires_resource('cpu')
    @threading_helper.requires_working_threading()
```

**Review by ZeroIntensity — COMMENTED** [review]
Some comments. I'll yield to you on whether or not it's worth trying to do this locklessly :)

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c:1412]: Good catch!

**Comment by ericsnowcurrently** [comment]
> * Should we backport this? I'm concerned we might make debuggers unhappy on 3.13

Let's circle back to this later.  My initial thought is that backporting would make sense.  What do you mean about making debuggers unhappy?

> * We should consider a lock-free algorithm for the freelist to help ease the contention if a lot of subintepreters are trying to get created.

Perhaps.  I don't anticipate there being enough simultaneous attempts that the overhead of any contention would matter.

> * If I understand correctly, this only applies to the preallocated initial thread. Why not make all thread states participate in the freelist?

I wanted to start off with this.  I'm not convinced that an actual freelist of thread states is worth it.  Conceptually it made sense as a solution for the pre-allocated thread state.  I debated on having a dedicated field for just that one vs. do the freelist thing.  Clearly I went with the latter, but now I'm thinking the illusion of a freelist isn't the best thing.  I may switch it back.

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c]: It may be.  If I switch back to a dedicated field for just the one thread state, instead of a freelist, then I'll definitely go with the atomic compare-and-swap.

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c:1429]: It resets it in `alloc_threadstate()` only if it is allocated there.  You may be right about keeping the old data being helpful while debugging, but I'm not sure it would make much of a difference.  Regardless, I'll think about it.  Thanks for bringing this up.

File: Lib/test/test_interpreters/test_stress.py
Comment (ericsnowcurrently) [reply Lib/test/test_interpreters/test_stress.py:36]: We get the same effect through other means. like, `import_helper.import_module()`, but there's no harm in being explicit about it.

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c:1412]: fixed

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c]: done

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c]: We really only need `_Py_atomic_store_ptr_relaxed()` here.  For an actual freelist we would need something more like what you have, but for now it only ever holds the one pointer.

Comment (ericsnowcurrently) [reply Python/pystate.c]: done

File: Lib/test/test_interpreters/test_stress.py
Comment (ericsnowcurrently) [reply Lib/test/test_interpreters/test_stress.py:36]: done

**Comment by ZeroIntensity** [comment]
> What do you mean about making debuggers unhappy?

Some low-level debuggers and profilers (something like PyStack, py-spy, and probably some others that I don't know about) might get a nasty surprise if we change where the main thread is stored. I don't think it's *that* important, as they're already aware of the maintenance burden that comes with relying on private implementation details, but it's not too convincing for a backport. (It's probably easier for downstream if we backport GH-126915, and then just put this on main for going forward, but I'm definitely biased :smile:)

> Perhaps. I don't anticipate there being enough simultaneous attempts that the overhead of any contention would matter.

I'm not too worried about the subinterpreter case, but instead for just normal multithreading under one interpreter. This will add additional overhead to calling something like `PyGILState_Ensure` with a lot of threads, and users might not be happy with how it scales.

File: Lib/test/test_interpreters/test_stress.py
Comment (ZeroIntensity) [reply Lib/test/test_interpreters/test_stress.py:36]: Good to know :)

File: Python/pystate.c
Comment (ericsnowcurrently) [reply Python/pystate.c:1429]: I'm going to hold off for now on calling `reset_threadstate()` in `alloc_threadstate()`, instead of in `free_threadstate()`, for the preallocated one.  The threadstate is statically initialized, so when `alloc_threadstate()` is called we know it is already ready to go.  By resetting in `free_threadstate()`, we ensure that condition holds.  If this proves to be pain for debugging we can adjust then, but for now I'd like to stick with the way I have it.

**Review by ZeroIntensity — COMMENTED** [review]
Ok, it looks like you've removed the freelist now. (I would change the title for the commit message.)

This looks pretty good, and I think this is small enough to not worry about giving debuggers too much of a headache. I'll do one final pass a little later today once you've figured out what you want to do about resets and then let's merge :)

**Comment by ericsnowcurrently** [comment]
> might get a nasty surprise if we change where the main thread is stored.

It is still found at `PyInterpreterState._initial_thread`, so it shouldn't be a problem.

> This will add additional overhead to calling something like PyGILState_Ensure with a lot of threads, and users might not be happy with how it scales.

I'd be surprised if lock/atomic contention here were more than insignificant relative to any of the other operations at play.  Also keep in mind that `PyMutex` is mostly a user-space lock, so some of the overhead of contention is reduced (relative to a kernel-provided lock).

**Comment by ZeroIntensity** [comment]
> It is still found at `PyInterpreterState._initial_thread`, so it shouldn't be a problem.

I noticed! I like this fix better than mine with the extra field, it's clever. In fact, we could reuse the `preallocated` field as a makeshift freelist for later; heap-allocated thread states could also get stored in that field, and then `alloc_threadstate` will use them just fine. 

> Also keep in mind that `PyMutex` is mostly a user-space lock, so some of the overhead of contention is reduced (relative to a kernel-provided lock).

Yeah, contention on those functions is probably not great right now anyways because of the `HEAD_LOCK`. I'm more talking generally, and trying to think ahead about what we should do in the future.

**Review by ZeroIntensity — APPROVED** [review]
LGTM

**Comment by ericsnowcurrently** [comment]
I'll add the backport to 3.13 label once gh-126995 is merged.  Otherwise sorting out the ABI data file becomes a much bigger pain.

**Comment by ericsnowcurrently** [comment]
I'm fairly sure it's unrelated.

**Pull request merged**
